pkg/config/structure/unmarshal.go

```diff
@@ -0,0 +1,368 @@
+				keys = append(keys, kstr)
+			}
+		}
+		return keys
+	}
+	return nil
+}
+
+// AsScalar returns the node as a leaf if possible, otherwise nil
+func (n *nodeimpl) AsScalar() leaf {
+	if n.typ != scalarNodeType {
+		return nil
+	}
+	return &leafimpl{val: n.val}
+}
+
+// AsList returns the node as a list if possible, otherwise nil
+func (n *nodeimpl) AsList() nodeList {
+	if n.typ != listNodeType {
+		return nil
+	}
+	return &nodeListImpl{val: n.val}
+}
+
+// Size returns the number of elements in the list
+func (n *nodeListImpl) Size() int {
+	return n.val.Len()
+}
+
+// Index returns the kth element of the list
+func (n *nodeListImpl) Index(k int) node {
+	elem := n.val.Index(k)
+	if elem.Kind() == reflect.Interface {
+		elem = elem.Elem()
+	}
+	return newNode(elem)
+}
+
+// GetBool returns the scalar as a bool, or an error otherwise
+func (n *leafimpl) GetBool() (bool, error) {
+	if n.val.Kind() == reflect.Bool {
+		return n.val.Bool(), nil
+	}
+	if n.val.Kind() == reflect.String {
+		if n.val.String() == "true" {
+			return true, nil
+		} else if n.val.String() == "false" {
+			return false, nil
+		}
+	}
+	return false, conversionError(n.val, "bool")
+}
+
+// GetInt returns the scalar as a int, or an error otherwise
+func (n *leafimpl) GetInt() (int, error) {
+	switch n.val.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return int(n.val.Int()), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return int(n.val.Uint()), nil
+	}
+	return 0, conversionError(n.val, "int")
+}
+
+// GetString returns the scalar as a string, or an error otherwise
+func (n *leafimpl) GetString() (string, error) {
+	if n.val.Kind() == reflect.String {
+		return n.val.String(), nil
+	}
+	return "", conversionError(n.val, "string")
+}
+
+func copyStruct(target reflect.Value, source node) error {
+	targetType := target.Type()
+	for i := 0; i < targetType.NumField(); i++ {
+		f := targetType.Field(i)
+		ch, _ := utf8.DecodeRuneInString(f.Name)
+		if unicode.IsLower(ch) {
+			continue
+		}
+		key := f.Name
+		if replace := f.Tag.Get("yaml"); replace != "" {
+			key = replace
+		} else if replace := f.Tag.Get("json"); replace != "" {
+			key = replace
+		} else if replace := f.Tag.Get("mapstructure"); replace != "" {
+			key = replace
+		}
+		child := source.GetChild(key)
+		if child == nil {
+			continue
+		}
+		err := copyAny(target.FieldByName(f.Name), child)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func copyMap(target reflect.Value, source node) error {
+	// TODO: Should handle maps with more complex types
+	ktype := reflect.TypeOf("")
+	vtype := reflect.TypeOf("")
+	mtype := reflect.MapOf(ktype, vtype)
+	results := reflect.MakeMap(mtype)
+
+	mapKeys := source.ChildrenKeys()
+	for _, mkey := range mapKeys {
+		child := source.GetChild(mkey)
+		if child == nil {
+			continue
+		}
+		scalar := child.AsScalar()
+		if scalar != nil {
+			mval, _ := scalar.GetString()
```

Comment:
If understand correctly, `GetString` would return an empty string if leaf can not get the string value. Do we want to store empty string as map values?

Reply:
Changed this call to `GetString` so that we check the error return. This call isn't quite a conversion: the var `scalar` represents a leaf node, and this call is getting the string value of that leaf node.